libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -0,0 +1,1074 @@
+				}
+			)");
+			templ("functionName", functionName);
+			templ("arrayDataSlot", arrayDataAreaFunction(_from));
+			return templ.render();
+		}
+		else
+		{
+			// Multiple items per slot
+			solAssert(_from.baseType()->storageBytes() <= 16, "");
+			solAssert(!_from.baseType()->isDynamicallyEncoded(), "");
+			solAssert(_from.baseType()->isValueType(), "");
+			bool dynamic = _to.isDynamicallyEncoded();
+			size_t storageBytes = _from.baseType()->storageBytes();
+			size_t itemsPerSlot = 32 / storageBytes;
+			// This always writes full slot contents to memory, which might be
+			// more than desired, i.e. it writes beyond the end of memory.
+			Whiskers templ(
+				R"(
+					function <functionName>(value, pos) <return> {
+						let length := <lengthFun>(value)
+						<storeLength> // might update pos
+						let originalPos := pos
+						let srcPtr := <dataArea>(value)
+						for { let i := 0 } lt(i, length) { i := add(i, <itemsPerSlot>) }
+						{
+							let data := sload(srcPtr)
+							<#items>
+								<encodeToMemoryFun>(<shiftRightFun>(data), pos)
+								pos := add(pos, <elementEncodedSize>)
+							</items>
+							srcPtr := add(srcPtr, 1)
+						}
+						pos := add(originalPos, mul(length, <elementEncodedSize>))
+						<assignEnd>
+					}
+				)"
+			);
+			templ("functionName", functionName);
+			templ("return", dynamic ? " -> end " : "");
+			templ("assignEnd", dynamic ? "end := pos" : "");
+			templ("lengthFun", arrayLengthFunction(_from));
+			if (_to.isDynamicallySized())
+				templ("storeLength", "mstore(pos, length) pos := add(pos, 0x20)");
+			else
+				templ("storeLength", "");
+			templ("dataArea", arrayDataAreaFunction(_from));
+			templ("itemsPerSlot", to_string(itemsPerSlot));
+			string elementEncodedSize = toCompactHexWithPrefix(_to.baseType()->calldataEncodedSize());
+			templ("elementEncodedSize", elementEncodedSize);
+			string encodeToMemoryFun = abiEncodingFunction(
+				*_from.baseType(),
+				*_to.baseType(),
+				_encodeAsLibraryTypes,
+				true
+			);
+			templ("encodeToMemoryFun", encodeToMemoryFun);
+			std::vector<std::map<std::string, std::string>> items(itemsPerSlot);
+			for (size_t i = 0; i < itemsPerSlot; ++i)
+				items[i]["shiftRightFun"] = shiftRightFunction(i * storageBytes * 8, false);
+			templ("items", items);
+			return templ.render();
+		}
+	});
+}
+
+string ABIFunctions::abiEncodingFunctionStringLiteral(
+	Type const& _from,
+	Type const& _to,
+	bool _encodeAsLibraryTypes
+)
+{
+	solAssert(_from.category() == Type::Category::StringLiteral, "");
+
+	string functionName =
+		"abi_encode_" +
+		_from.identifier() +
+		"_to_" +
+		_to.identifier() +
+		(_encodeAsLibraryTypes ? "_lib" : "");
+	return createFunction(functionName, [&]() {
+		auto const& strType = dynamic_cast<StringLiteralType const&>(_from);
+		string const& value = strType.value();
+		solAssert(_from.sizeOnStack() == 0, "");
+
+		if (_to.isDynamicallySized())
+		{
+			Whiskers templ(R"(
+				function <functionName>(pos) -> end {
```

Comment:
It returns the "new memory pointer" if and only if it is dynamically sized. You can take a look at how the functions are called, it is consistent and should be checked by the julia analyzer.